Allow for <input type=checkbox switch> haptic feedback on click

annevk opened the pull request:
#### a81bf20129ac80f7b807079fe7e5323eecb3e4df
<pre>
Allow for &lt;input type=checkbox switch&gt; haptic feedback on click
<a href="https://bugs.webkit.org/show_bug.cgi?id=271711">https://bugs.webkit.org/show_bug.cgi?id=271711</a>

Reviewed by Aditya Keerthi.

And enable this on HAVE(UI_IMPACT_FEEDBACK_GENERATOR) platforms. For
macOS there should no functional change.

(Not guarding the click path on macOS appears to have the same end
result as well as the haptic cannot be observed, but that does not seem
optimal as more code ends up being executed.)

* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/html/CheckboxInputType.cpp:
(WebCore::CheckboxInputType::willDispatchClick):
(WebCore::switchAnimationUpdateInterval):
(WebCore::CheckboxInputType::performSwitchVisuallyOnAnimation):
(WebCore::CheckboxInputType::updateIsSwitchVisuallyOnFromAbsoluteLocation):
* Source/WebCore/html/CheckboxInputType.h:
* Source/WebCore/html/SwitchTrigger.h: Added.
* Source/WebCore/rendering/RenderTheme.h:
(WebCore::RenderTheme::hasSwitchHapticFeedback const):
* Source/WebCore/rendering/ios/RenderThemeIOS.h:
* Source/WebCore/rendering/mac/RenderThemeMac.h:

Canonical link: <a href="https://commits.webkit.org/276727@main">https://commits.webkit.org/276727@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a4f3f6e7e85cdb190f7b658c01e39b84df402ca2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/45502 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/24628 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/48040 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/48169 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/41511 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/47809 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/28877 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/22022 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/37290 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/46080 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/21713 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/39263 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/18414 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/19111 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/40350 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/3548 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/41822 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/40677 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/49902 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/20490 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/17019 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/44365 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/21796 "Built successfully") | | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/43197 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10115 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/22156 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/21484 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/html/CheckboxInputType.cpp:
```diff
@@ -364,6 +364,15 @@ void CheckboxInputType::performSwitchAnimation(SwitchAnimationType type)
+    performSwitchAnimation(SwitchAnimationType::VisuallyOn);
+    if (!RenderTheme::singleton().hasSwitchHapticFeedback(trigger))
+        return;
+    if (auto* page = element()->document().page())
```
```suggestion
    if (RefPtr page = element()->document().page())
```

annevk committed:
Allow for <input type=checkbox switch> haptic feedback on click
https://bugs.webkit.org/show_bug.cgi?id=271711

Reviewed by Aditya Keerthi.

And enable this on HAVE(UI_IMPACT_FEEDBACK_GENERATOR) platforms. For
macOS there should no functional change.

(Not guarding the click path on macOS appears to have the same end
result as well as the haptic cannot be observed, but that does not seem
optimal as more code ends up being executed.)

* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/html/CheckboxInputType.cpp:
(WebCore::CheckboxInputType::willDispatchClick):
(WebCore::switchAnimationUpdateInterval):
(WebCore::CheckboxInputType::performSwitchVisuallyOnAnimation):
(WebCore::CheckboxInputType::updateIsSwitchVisuallyOnFromAbsoluteLocation):
* Source/WebCore/html/CheckboxInputType.h:
* Source/WebCore/html/SwitchTrigger.h: Added.
* Source/WebCore/rendering/RenderTheme.h:
(WebCore::RenderTheme::hasSwitchHapticFeedback const):
* Source/WebCore/rendering/ios/RenderThemeIOS.h:
* Source/WebCore/rendering/mac/RenderThemeMac.h:

Canonical link: https://commits.webkit.org/276727@main
